substrate/frame/nomination-pools/src/lib.rs

```diff
@@ -3263,31 +3349,40 @@ impl<T: Config> Pallet<T> {
 }
 
 impl<T: Config> sp_staking::OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// Reduces the balances of the [`SubPools::with_era`], that belong to the pool involved in the
+	/// slash, to the amount that is defined in the `slashed_unlocking` field of
+	/// [`sp_staking::OnStakingUpdate::on_slash`]
+	///
+	/// Emits [`Event::<T>::PoolsSlashed`].
 	fn on_slash(
 		pool_account: &T::AccountId,
 		// Bonded balance is always read directly from staking, therefore we don't need to update
 		// anything here.
 		slashed_bonded: BalanceOf<T>,
 		slashed_unlocking: &BTreeMap<EraIndex, BalanceOf<T>>,
+		total_slashed: BalanceOf<T>,
 	) {
-		if let Some(pool_id) = ReversePoolIdLookup::<T>::get(pool_account) {
-			let mut sub_pools = match SubPoolsStorage::<T>::get(pool_id).defensive() {
-				Some(sub_pools) => sub_pools,
-				None => return,
-			};
-			for (era, slashed_balance) in slashed_unlocking.iter() {
-				if let Some(pool) = sub_pools.with_era.get_mut(era) {
+		let Some(pool_id) = ReversePoolIdLookup::<T>::get(pool_account) else { return };
+		// As the slashed account belongs to a `BondedPool` the `TotalValueLocked` decreases and
+		// an event is emitted.
+		TotalValueLocked::<T>::mutate(|tvl| {
+			tvl.saturating_reduce(total_slashed);
+		});
+
+		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id) {
+			// set the reduced balance for each of the `SubPools`
+			slashed_unlocking.iter().for_each(|(era, slashed_balance)| {
+				if let Some(pool) = sub_pools.with_era.get_mut(era).defensive() {
```

Comment:
Some of the slashed balance might be from an era that's been merged into no_era right? I am wondering if the older code was correct as well. If era is not found in a subpool with eras, the remaining non-slashed balance should be at the end slashed from subpool with no_era (I think). cc: @kianenigma @gpestana

Reply:
I think a merge into a no_era pool is possible to minimize the storage taken by pool users that unbond but just never withdraw. If they would keep their own pool for every withdraw the storage gets bloated pretty quickly. So these unwanted pools get screwed in terms of having their own "point to balance ratio" and just get merged into this one big `SubPool::no_era` but only if they are way past the actual unbonding duration.

Furthermore in the current doc [this](https://github.com/paritytech/polkadot-sdk/blob/28aa4e94ea38a0647a97e4a77a61b08f02cf433e/substrate/frame/nomination-pools/src/lib.rs#L1537C1-L1539C95) explains why this should not be allowed to be not found. I guess the old code lacked the defensive error.

Reply:
tldr: 
> Some of the slashed balance might be from an era that's been merged into no_era right? 
I dont think so.

Reply:
@Ank4n I revisited this and stumbled across this [doc](https://github.com/paritytech/polkadot-sdk/blob/f1994c8690baa7f718f3fc5a351c8c6c9661e156/substrate/frame/staking/src/slashing.rs#L590C1-L593C28) comment in `slashing.rs`. 

The way I interpret this is that still `SubPools::no_era` will be untouched but any amount that remains of the slash might need to be deducted from the pool's `RewardPool`. This was neither accounted for in the previous code. 

Are these lines also where your concern originates from?